lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,768 @@
+  end
+
+  defp terminate_children(children, state) do
+    {pids, times, stacks} = monitor_children(children)
+    size = map_size(pids)
+
+    timers =
+      Enum.reduce(times, %{}, fn {time, pids}, acc ->
+        Map.put(acc, :erlang.start_timer(time, self(), :kill), pids)
+      end)
+
+    stacks = wait_children(pids, size, timers, stacks)
+
+    for {pid, {child, reason}} <- stacks do
+      report_error(:shutdown_error, reason, pid, child, state)
+    end
+
+    :ok
+  end
+
+  defp monitor_children(children) do
+    Enum.reduce(children, {%{}, %{}, %{}}, fn
+      {_, {:restarting, _}}, acc ->
+        acc
+
+      {pid, {_, restart, _, _, _} = child}, {pids, times, stacks} ->
+        case monitor_child(pid) do
+          :ok ->
+            times = exit_child(pid, child, times)
+            {Map.put(pids, pid, child), times, stacks}
+
+          {:error, :normal} when restart != :permanent ->
+            {pids, times, stacks}
+
+          {:error, reason} ->
+            {pids, times, Map.put(stacks, pid, {child, reason})}
+        end
+    end)
+  end
+
+  defp monitor_child(pid) do
+    ref = Process.monitor(pid)
+    Process.unlink(pid)
+
+    receive do
+      {:EXIT, ^pid, reason} ->
+        receive do
+          {:DOWN, ^ref, :process, ^pid, _} -> {:error, reason}
+        end
+    after
+      0 -> :ok
+    end
+  end
+
+  defp exit_child(pid, {_, _, shutdown, _, _}, times) do
+    case shutdown do
+      :brutal_kill ->
+        Process.exit(pid, :kill)
+        times
+
+      :infinity ->
+        Process.exit(pid, :shutdown)
+        times
+
+      time ->
+        Process.exit(pid, :shutdown)
+        Map.update(times, time, [pid], &[pid | &1])
+    end
+  end
+
+  defp wait_children(_pids, 0, timers, stacks) do
+    for {timer, _} <- timers do
+      _ = :erlang.cancel_timer(timer)
+
+      receive do
+        {:timeout, ^timer, :kill} -> :ok
+      after
+        0 -> :ok
+      end
+    end
+
+    stacks
+  end
+
+  defp wait_children(pids, size, timers, stacks) do
+    receive do
+      {:DOWN, _ref, :process, pid, reason} ->
+        case pids do
+          %{^pid => child} ->
+            stacks = wait_child(pid, child, reason, stacks)
+            wait_children(pids, size - 1, timers, stacks)
+
+          %{} ->
+            wait_children(pids, size, timers, stacks)
+        end
+
+      {:timeout, timer, :kill} ->
+        for pid <- Map.fetch!(timers, timer), do: Process.exit(pid, :kill)
+        wait_children(pids, size, Map.delete(timers, timer), stacks)
+    end
+  end
+
+  defp wait_child(pid, {_, _, :brutal_kill, _, _} = child, reason, stacks) do
+    case reason do
+      :killed -> stacks
+      _ -> Map.put(stacks, pid, {child, reason})
+    end
+  end
+
+  defp wait_child(pid, {_, restart, _, _, _} = child, reason, stacks) do
+    case reason do
+      :shutdown -> stacks
+      :normal when restart != :permanent -> stacks
+      reason -> Map.put(stacks, pid, {child, reason})
```

Comment:
For transient/temporary do we want to report `{:shutdown, _}`?

Reply:
Good catch.